java/semantickernel-api/src/main/java/com/microsoft/semantickernel/tokenizers/GPT3Tokenizer.java

```diff
@@ -0,0 +1,290 @@
+                    maxUtf8Length = utf8Length;
+                }
+            }
+
+            // Ensure we have a sufficient char array to accommodate maxUtf8Length chars.
+            // The byte-to-char mapping scheme employed is 1:1, so we'll end up needing 1 char
+            // for every 1 UTF8 byte.
+            char[] chars = new char[maxUtf8Length];
+
+            // Rather than using separate space for the UTF8 bytes, we just reinterpret the char
+            // array
+            // as a byte array. Since our mapping is 1:1, the space required for the bytes will
+            // always
+            // be half of the space required for the chars. We can UTF8-encode into the first half,
+            // and
+            // then walk backwards through the bytes, using the byte-to-char mapping scheme to
+            // populate
+            // the chars from the back to the front. By going in reverse, we guarantee we won't
+            // overwrite
+            // any bytes we haven't yet seen.
+            byte[] bytes = new byte[maxUtf8Length];
+
+            // Now that our space is created, do the actual encoding.
+            matcher.reset();
+            while (matcher.find()) {
+                String matchValue = matcher.group();
+
+                // Encode the UTF8 bytes.
+                int bytesWritten = encodingUtf8GetBytes(matchValue, bytes);
+
+                // Translate those bytes into chars.
+                for (int i = bytesWritten - 1; i >= 0; i--) {
+                    chars[i] = bytesToUnicode[bytes[i] & 0xFF];
+                }
+
+                // Evaluate the BPE for the encoded chars.
+                for (String encoding : bytePairEncoding(new String(chars, 0, bytesWritten))) {
+                    bpeTokens.add(GPT3Settings.encoder.get(encoding));
+                }
+            }
+        }
+
+        return bpeTokens;
+    }
+
+    public static List<Integer> encode(StringBuilder stringBuilder) {
+        if (stringBuilder != null) {
+            return encode(stringBuilder.toString());
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    public static List<Integer> encode(char[] chars) {
+        if (chars != null) {
+            return encode(new String(chars));
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    public static List<Integer> encode(Iterable<Character> chars) {
+        if (chars != null) {
+            StringBuilder sb = new StringBuilder();
+            for (Character c : chars) {
+                sb.append(c);
+            }
+            return encode(sb.toString());
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    private static int encodingUtf8GetByteCount(String str) {
+        return str.getBytes(StandardCharsets.UTF_8).length;
+    }
+
+    private static int encodingUtf8GetBytes(String str, byte[] bytes) {
+        byte[] strBytes = str.getBytes(StandardCharsets.UTF_8);
```

Comment:
My first reaction here was 'possible NPE'. But then I looked closer and see that the caller ensures str is not null and bytes is not null. I think this is a good place for an `assert str != null` and `assert bytes != null`. This is just my opinion but this tells the reader that you expect the caller to check for null. In other words, you are asserting that str is not null. The assert is only enabled with -ea.